### PR TITLE
Move email templates to separate model

### PR DESCRIPTION
**Pull request by maarcingebala**
Currently, email templates are loaded along with the plugin configuration for every request. Since they are HTML and may weigh much more than regular config values, having them loaded with every request may significantly affect the network bandwidth during high traffic. We optimize this by moving templates to a separate model and loading them only when they are needed, just before sending the email.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-5449 Plugin manager fetches several megabytes of data for every LUSH request](https://app.clickup.com/t/2549495/SALEOR-5449)

File: saleor/plugins/admin_email/plugin.py [hunk 1]
```diff
@@ -217,26 +219,53 @@ def __init__(self, *args, **kwargs):
             ],
         )
 
+    def resolve_plugin_configuration(
+        self, request
+    ) -> Union[PluginConfigurationType, Promise[PluginConfigurationType]]:
+        # Get email templates from the database and merge them with self.configuration.
+        configuration = super().resolve_plugin_configuration(request)
+        if not self.db_config:
+            return configuration
```
Comment (fowczarek) [inline saleor/plugins/admin_email/plugin.py:228]:
```suggestion
        if not self.db_config:
            configuration = super().resolve_plugin_configuration(request)
            return configuration
```

File: saleor/plugins/user_email/plugin.py [hunk 2]
```diff
@@ -410,29 +417,101 @@ def __init__(self, *args, **kwargs):
             ],
         )
 
+    def resolve_plugin_configuration(
+        self, request
+    ) -> Union[PluginConfigurationType, Promise[PluginConfigurationType]]:
+        # Get email templates from the database and merge them with self.configuration.
+        configuration = super().resolve_plugin_configuration(request)
+        if not self.db_config:
+            return configuration
```
Comment (fowczarek) [inline saleor/plugins/user_email/plugin.py:426]:
```suggestion
        if not self.db_config:
            configuration = super().resolve_plugin_configuration(request)
            return configuration
```

File: saleor/plugins/user_email/plugin.py [hunk 3]
```diff
@@ -410,29 +417,101 @@ def __init__(self, *args, **kwargs):
...
+            plugin_configuration.active = cleaned_data["active"]
+
+        cls.validate_plugin_configuration(
+            plugin_configuration, email_templates_data=email_templates_data
+        )
+        cls.pre_save_plugin_configuration(plugin_configuration)
+
+        plugin_configuration.save()
+
+        # TODO: move this logic to admin email as well
```
[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/plugins/user_email/plugin.py]: Yep, removed the comment.

File: saleor/graphql/plugins/dataloaders.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+from collections import defaultdict
+
+from ...plugins.models import EmailTemplate
+from ..core.dataloaders import DataLoader
+
+
+class EmailTemplatesByPluginConfigurationLoader(DataLoader):
+    """Loads pages by pages type ID."""
```
[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/graphql/plugins/dataloaders.py]: Fixed

File: saleor/plugins/migrations/0010_auto_20220104_1239.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+        EmailTemplate.objects.bulk_create(email_templates)
+
+        # After saving new EmailTemplate instances to DB, recreate
+        # PluginConfiguration.configuration list but with email templates omitted.
+        new_configuration = []
+        for json_config in plugin_config_obj.configuration:
+            config_key = json_config.get("name")
+            if config_key not in EMAIL_NAMES:
+                new_configuration.append(json_config)
```
Comment (fowczarek) [inline saleor/plugins/migrations/0010_auto_20220104_1239.py:64]: Maybe we could reverse it. Remove `EMAIL_NAMES` from the existing configuration and save the object?

File: saleor/plugins/migrations/0010_auto_20220104_1239.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+]
+
+
+def move_email_templates_to_separate_model(apps, schema):
+    PluginConfiguration = apps.get_model("plugins", "PluginConfiguration")
+    EmailTemplate = apps.get_model("plugins", "EmailTemplate")
+
+    plugin_configs = PluginConfiguration.objects.filter(identifier__in=EMAIL_PLUGINS)
+
+    for plugin_config_obj in plugin_configs:
```
Comment (fowczarek) [inline saleor/plugins/migrations/0010_auto_20220104_1239.py:41]:
```suggestion
    for plugin_config_obj in plugin_configs.iterator():
```

Comment (maarcingebala) [reply saleor/plugins/migrations/0010_auto_20220104_1239.py]: Fixed

File: saleor/plugins/user_email/plugin.py (hunk 2 above)
Comment (maarcingebala) [reply saleor/plugins/user_email/plugin.py]: Fixed

File: saleor/plugins/admin_email/plugin.py (hunk 1 above)
Comment (maarcingebala) [reply saleor/plugins/admin_email/plugin.py]: Fixed

File: saleor/graphql/plugins/dataloaders.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    def batch_load(self, keys):
+        email_templates = EmailTemplate.objects.using(
+            self.database_connection_name
+        ).filter(plugin_configuration_id__in=keys)
+
+        config_to_template = defaultdict(list)
+        for et in email_templates:
+            config_to_template[et.plugin_configuration_id].append(et)
+
+        return [config_to_template[key] for key in keys]
```
[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/graphql/plugins/dataloaders.py]: Fixed

File: saleor/plugins/migrations/0010_auto_20220104_1239.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+        # After saving new EmailTemplate instances to DB, recreate
+        # PluginConfiguration.configuration list but with email templates omitted.
+        new_configuration = []
+        for json_config in plugin_config_obj.configuration:
+            config_key = json_config.get("name")
+            if config_key not in EMAIL_NAMES:
+                new_configuration.append(json_config)
+
+        plugin_config_obj.configuration = new_configuration
```
[the remark this answers is not in the document]
Comment (korycins) [reply saleor/plugins/migrations/0010_auto_20220104_1239.py:66]: Ignore this. I had inccorect assumption based on different plugin. Sorry!

[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/plugins/migrations/0010_auto_20220104_1239.py:66]: No worries Maciej!

**Review by fowczarek — APPROVED** [review]

**Review by korycins — APPROVED** [review]

File: saleor/plugins/manager.py [hunk 9]
```diff
@@ -111,6 +114,21 @@ def __init__(self, plugins: List[str]):
             for channel in channels:
                 self.plugins_per_channel[channel.slug].extend(self.global_plugins)
 
+    def _get_db_plugin_configs(self):
+        with opentracing.global_tracer().start_active_span("_get_db_plugin_configs"):
+            qs = PluginConfiguration.objects.all().prefetch_related("channel")
```
Comment (NyanKiyoshi) [inline saleor/plugins/manager.py:119]: We might want to use read-replicas here

Comment (maarcingebala) [reply saleor/plugins/manager.py]: Thanks, fixed.

**Review by cmiacz — APPROVED** [review]

File: saleor/plugins/user_email/plugin.py [hunk 10]
```diff
@@ -410,29 +417,99 @@ def __init__(self, *args, **kwargs):
...
+                EmailTemplate.objects.update_or_create(
+                    name=et_data["name"],
+                    plugin_configuration=plugin_configuration,
+                    defaults={"value": et_data["value"]},
+                )
+
+        if plugin_configuration.configuration:
+            # Let's add a translated descriptions and labels
+            cls._append_config_structure(plugin_configuration.configuration)
+        return plugin_configuration
```
Comment (IKarbowiak) [inline saleor/plugins/user_email/plugin.py:515]: It's not the same as in the `saleor/plugins/admin_email/plugin.py` `save_plugin_configuration` method? Maybe we can create utils from this?

**Review by IKarbowiak — APPROVED** [review]

File: saleor/plugins/user_email/plugin.py (hunk 10 above)
Comment (maarcingebala) [reply saleor/plugins/user_email/plugin.py:527]: Thanks for catching this anyway :)

**Pull request merged**